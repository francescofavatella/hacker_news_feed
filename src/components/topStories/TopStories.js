import React, { Component } from 'react';
import axios from 'axios';
import InfiniteScroll from 'react-infinite-scroller';

import styles from './TopStories.module.css';
import { StoryContainer } from '../storyContainer/StoryContainer';

export class TopStories extends Component {
  constructor() {
    super();
    this.state = {
      stories: [],
      page: 0,
      storiesForPage: 10,
    };
  }

  componentDidMount() {
    //TODO: refactor this code for testing purpose
    axios
      .get('https://hacker-news.firebaseio.com/v0/topstories.json')
      .then(res => {
        const data = res.data;
        this.setState((state, props) => ({
          stories: data,
        }));
      })
      // TODO: handle error during loading
      .catch(err => console.log(err));
  }

  loadFunc = page => {
    this.setState((state, props) => ({
      page: page,
    }));
  };

  hasMore = () => {
    const { stories, page, storiesForPage } = this.state;
    return stories.length > page * storiesForPage;
  };

  render() {
    const { stories, page, storiesForPage } = this.state;
    return (
      <InfiniteScroll
        pageStart={0}
        loadMore={this.loadFunc}
        hasMore={this.hasMore()}
        loader={
          <div className="loader" key={0}>
            Loading ...
          </div>
        }
      >
        <ul className={styles.container}>
          {stories
            .slice(0, Math.min(page * storiesForPage, stories.length))
            .map(storyId => (
              <StoryContainer key={storyId} storyId={storyId} />
            ))}
        </ul>
      </InfiniteScroll>
    );
  }
}

export default TopStories;
