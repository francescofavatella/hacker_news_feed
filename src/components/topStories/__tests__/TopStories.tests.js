import React from 'react';
import TestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import TopStories from '../TopStories';

describe('TopStories', () => {
  describe('with no props', () => {
    const component = <TopStories />;

    it('renders without crashing', () => {
      shallow(component);
    });

    it('matches snapshot', () => {
      const tree = TestRenderer.create(component).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
