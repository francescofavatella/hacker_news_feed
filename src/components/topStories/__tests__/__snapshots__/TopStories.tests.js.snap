// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopStories with no props matches snapshot 1`] = `
<div>
  <ul
    className="container"
  />
</div>
`;
